controllers/priorityexpander_controller.go

```diff
@@ -0,0 +1,240 @@
+		// we'll ignore not-found errors, since they can't be fixed by an immediate
+		// requeue (we'll need to wait for a new notification), and we can get them
+		// on deleted requests.
+		err = client.IgnoreNotFound(err)
+		if err != nil {
+			log.Error(err, "Unable to fetch PriorityExpander")
+		}
+
+		return ctrl.Result{}, err
+	}
+
+	// Fetch clusterautoscaler status configmap.
+	var cm kcore_v1.ConfigMap
+	if err := r.Get(ctx, types.NamespacedName{
+		Namespace: r.Configuration.ClusterAutoscalerStatusNamespace,
+		Name:      r.Configuration.ClusterAutoscalerStatusName,
+	}, &cm); err != nil {
+		log.Error(
+			err,
+			"Unable to fetch ClusterAutoscaler status configmap",
+			"namespace", r.Configuration.ClusterAutoscalerStatusNamespace,
+			"name", r.Configuration.ClusterAutoscalerStatusName,
+		)
+
+		return ctrl.Result{}, err
+	}
+	// Get human readable status from configmap...
+	readableStatus, ok := cm.Data["status"]
+	if !ok {
+		err := fmt.Errorf("invalid configmap: no status")
+		log.Error(
+			err,
+			"unable to parse ClusterAutoscaler status",
+			"namespace", r.Configuration.ClusterAutoscalerStatusNamespace,
+			"name", r.Configuration.ClusterAutoscalerStatusName,
+		)
+	}
+
+	// ... and parse it.
+	status := clusterautoscaler.ParseReadableString(readableStatus)
+
+	var oroot = map[string]map[string]int32{}
+	for _, node := range status.NodeGroups {
+		oroot[node.Name] = make(map[string]int32)
+		oroot[node.Name]["CloudProviderTarget"] = node.Health.CloudProviderTarget
+		oroot[node.Name]["Ready"] = node.Health.Ready
+		oroot[node.Name]["Unready"] = node.Health.Unready
+		oroot[node.Name]["NotStarted"] = node.Health.NotStarted
+		oroot[node.Name]["LongNotStarted"] = node.Health.LongNotStarted
+		oroot[node.Name]["Registered"] = node.Health.Registered
+		oroot[node.Name]["LongUnregistered"] = node.Health.LongUnregistered
+		oroot[node.Name]["MinSize"] = node.Health.MinSize
+		oroot[node.Name]["MaxSize"] = node.Health.MaxSize
+	}
+
+	//Create new PriorityExpander template and parse it
+	t, err := template.New("template").Parse(pexp.Spec.Template)
+	if err != nil {
+		log.Error(
+			err,
+			"Error parsing PriorityExpander template. Check your syntax and/or rtfm.",
+		)
+		return ctrl.Result{}, err
+	}
+
+	buf := new(bytes.Buffer)
+	if err := t.Execute(buf, oroot); err != nil {
+		log.Error(
+			err,
+			"Error parsing generating template output.",
+		)
+		return ctrl.Result{}, err
+	}
+
+	// parsed content: fmt.Println(buf.String())
+	// Create the new ConfigMap object
+	pecm := kcore_v1.ConfigMap{
+		TypeMeta: kmeta_v1.TypeMeta{
+			Kind:       "ConfigMap",
+			APIVersion: "v1",
+		},
+		ObjectMeta: kmeta_v1.ObjectMeta{
+			Name:      r.Configuration.ClusterAutoscalerPEConfigMapName,
+			Namespace: r.Configuration.ClusterAutoscalerStatusNamespace,
+		},
+		Data: map[string]string{
+			"priorities": buf.String(),
+		},
+	}
+
+	// Update it inplace. The ConfigMap _must_ exist.
+	if err := r.Update(ctx, &pecm); err != nil {
```

Comment:
You should use controller-runtime `CreateOrUpdate` method instead to cover the case where the configmap does not exist yet.

Reply:
CreateOrUpdate does not exist.

Reply:
Ok, it does exist. It seems it brings quite some complexity to the code, regarding what's done on quortex-operator. Wondering if I should rather stick to the Client library and check if ConfigMap exists or not.
Or, maybe we should consider shipping the default ConfigMap with the chart.

Reply:
It can be simple : 
```
	pecm := &kcore_v1.ConfigMap{
		ObjectMeta: kmeta_v1.ObjectMeta{
			Name:      r.Configuration.ClusterAutoscalerPEConfigMapName,
			Namespace: r.Configuration.ClusterAutoscalerStatusNamespace,
		},
	}

	res, err := ctrl.CreateOrUpdate(ctx, r.Client, pecm, func() error {
		cm.Data = map[string]string{
			"priorities": buf.String(),
		}
	})
	if err != nil {
		log.Error(
			err,
			"Unable to update ClusterAutoscaler priority expander configmap",
			"namespace", r.Configuration.ClusterAutoscalerStatusNamespace,
			"name", r.Configuration.ClusterAutoscalerStatusName,
		)

		return ctrl.Result{}, err
	}
```